### ISSUE TITLE
Cannot adapt stack size for functions called during production of iterators

### ISSUE THREAD

**Issue by 01mf02** [rayon-rs/rayon#900]
I have an iterator that produces its elements by calling a recursive function (`rec`).
When operating in parallel on the iterator (via Rayon's `par_bridge`), the call to the recursive function results in a stack overflow,
whereas calling the recursive function outside of the iterator, there is no stack overflow.
I thus conclude that Rayon limits the stack size of the function that is called when producing the iterator.
However, I have not found a way to increase this limit --- calling `stack_size` on the `ThreadPoolBuilder` does not change the behaviour.

An example that exposes the behaviour in Rayon 1.5.1 ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e88ad3a3af349a27e951ef4f91b02351)):

~~~ rust
use rayon::prelude::*;
use std::iter;

fn rec(n: usize) -> usize {
    if n == 0 {
        0
    } else {
        1 + rec(n - 1)
    }
}

const ITERATIONS: usize = 100_000;
const STACK_SIZE: usize = ITERATIONS * 4;

fn main() {
    // when setting ITERATIONS to a higher value (say 200_000),
    // the next line produces a stack overflow as well, so recursion is definitely happening
    println!("Normal output is: {}", rec(ITERATIONS));

    let out = iter::once_with(|| rec(ITERATIONS)).all(|x| x > 0);
    println!("Iteration output is: {}", out);

    // doing the following does not seem to have an effect
    rayon::ThreadPoolBuilder::new()
        .stack_size(STACK_SIZE)
        .build_global()
        .unwrap();

    // ----> the next line overflows the stack! <----
    let out = iter::once_with(|| rec(ITERATIONS)).par_bridge().all(|x| x > 0);
    println!("Rayon output is: {}", out);
}
~~~

(In this example, when compiling in release mode, the error goes away because the call to `rec` is precomputed, but I have reproduced the problem in my application also in release mode.)

Is there a way to increase the stack size for functions that are called during the generation of an iterator?

**Comment by cuviper** [comment]
You're actually making it smaller with `STACK_SIZE = 100_000 * 4`.
The default thread stack size is 2 MiB, as documented in [`std::thread`](https://doc.rust-lang.org/stable/std/thread/index.html#stack-size).

In case it's not clear, note that the stack size is not directly _how many_ nested calls are possible, but rather how much local space in bytes is available for all of those calls. For instance, x86_64 needs at least 8 bytes per call just to push and pop the instruction pointer, and you'll need more for whatever local data is used in that call. Debug mode is quite wasteful with stack space too.

I would generally consider it bad form to recurse so deeply though -- 100k is a lot. I'd suggest converting your code to a loop if possible -- the optimizer can sometimes do if it's tail-call recursion, but that's hard to rely on. Sometimes a local `loop` with your own `Vec` "stack" can get the job done too, even with more complicated recursion patterns.

**Comment by 01mf02** [comment]
Ah, this clears things up! Thank you for your response!

For posterity: I tried setting the stack size in Rayon to the output of `ulimit -s`. However, I ignored that `ulimit` returns the stack size in *kilo*bytes, whereas Rayon takes it in bytes.

On the topic of recursing in general: The recursion happens in the parser of my application, where it is hard to replace it with a `loop`. Only on some (pathologically nested) input, there was so much recursion as to overflow the stack.
